src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -0,0 +1,287 @@
+                var type = typeElementToType(varOp.varValueType());
+                var v = new Symbol.VarSymbol(LocalVarFlags, name, type, ms);
+                yield treeMaker.VarDef(v, (JCTree.JCExpression) valueToTree.get(varOp.initOperand()));
+            }
+            case CoreOp.VarAccessOp.VarLoadOp varLoadOp
+                    when varLoadOp.varOp().initOperand() instanceof Block.Parameter p2 -> valueToTree.get(p2);
+            case CoreOp.VarAccessOp.VarLoadOp varLoadOp ->
+                    treeMaker.Ident((JCTree.JCVariableDecl) valueToTree.get(varLoadOp.varOperand()));
+            case CoreOp.FieldAccessOp.FieldLoadOp fieldLoadOp -> {
+                // Type.fieldName
+                // if instance field we will use the same thechnique as in invokeOpToTree
+                int flags;
+                try {
+                    flags = fieldLoadOp.fieldDescriptor().resolveToMember(MethodHandles.lookup()).getModifiers();
+                } catch (ReflectiveOperationException e) {
+                    throw new RuntimeException(e);
+                }
+                var name = names.fromString(fieldLoadOp.fieldDescriptor().name());
+                var type = typeElementToType(fieldLoadOp.resultType());
+                var owner = typeElementToType(fieldLoadOp.fieldDescriptor().refType());
+                var sym = new Symbol.VarSymbol(flags, name, type, owner.tsym);
+                yield treeMaker.Select(treeMaker.Ident(owner.tsym), sym);
+            }
+            case CoreOp.ArrayAccessOp.ArrayStoreOp arrayStoreOp -> {
+                var array = arrayStoreOp.operands().get(0);
+                var val = arrayStoreOp.operands().get(1);
+                var index = arrayStoreOp.operands().get(2);
+                var as = treeMaker.Assign(
+                        treeMaker.Indexed((JCTree.JCExpression) valueToTree.get(array), (JCTree.JCExpression) valueToTree.get(index)),
+                        (JCTree.JCExpression) valueToTree.get(val)
+                );
+                yield treeMaker.Exec(as);
+                // body builder are created but never passed when creating the op, why ?
+            }
+            default -> throw new IllegalStateException("Op -> JCTree not supported for :" + op.getClass().getName());
+        };
+        valueToTree.put(op.result(), tree);
+        return tree;
+    }
+
+    public JCTree.JCMethodDecl transformFuncOpToAST(CoreOp.FuncOp funcOp, Name methodName) {
+        Assert.check(funcOp.body().blocks().size() == 1);
+
+        var paramTypes = List.of(crSym.opFactoryType, crSym.typeElementFactoryType);
+        var mt = new Type.MethodType(paramTypes, crSym.opType, List.nil(), syms.methodClass);
+        ms = new Symbol.MethodSymbol(PUBLIC | STATIC | SYNTHETIC, methodName, mt, currClassSym);
+        currClassSym.members().enter(ms);
+
+        // TODO add VarOps in OpBuilder
+        funcOp = addVarsWhenNecessary(funcOp);
+        funcOp.writeTo(System.out);
```

Comment:
This should be removed?

Reply:
`addVarsWhenNecessary` is used to add necessary VarOp in the code model before transformation to AST. For example if an op result is used more than once, we introduce a VarOp to hold the value, because that's what the equivalent Java code will have. This makes the transformation from code model to AST a bit easier.
We also use it to in case we want an Op (e.g. InvokeOp) to be appended right away to the method we are building. As you now the block of a method is a list of statements, some operations maps to an expression, so they won't be appended until needed.

Reply:
Normally, the OpBuilder should produces a model that contains the VarOp when they should be. But I decided to have a seperate transformation as a quick fix. Later I will change the OpBuilder.

Reply:
> But I decided to have a seperate transformation as a quick fix. Later I will change the OpBuilder.

As stated in another comment, I'm not sure changing the op builder is needed -- as the javac code can easily workaround what the op builder does (e.g. what it does doesn't seem obviously wrong to me). But if the decision is to fix op builder that's fine too

Reply:
It would be nice to directly avoid changing OpBuilder (investing in a SSA to Var transformation, using liveness analysis, is probably the general right approach), but there might be a more focused approach to generating the AST so we can avoid this.

Reply:
Thanks for confirming @PaulSandoz -- we can defer this improvement to a follow up change -- and keep the var-adding transformation in this PR for now.